### PR TITLE
[Fix-16900][api] fix when the copied workflow, the logical node switch loses the flow relationship

### LINKED ISSUE #16900: [Bug]  In the copied workflow, the logical node switch loses the flow relationship
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

当去复制一个拥有逻辑节点switch的工作流时，复制的工作流中逻辑节点的流转关系丢失，需要手动进去配置流转关系。如果量少还能接受，但当工作流成百上千时，就比较麻烦了。

### What you expected to happen

前端展示存在问题

### How to reproduce

当去复制一个拥有逻辑节点switch的工作流时，复制的工作流中逻辑节点的流转关系丢失，需要手动进去配置流转关系。如果量少还能接受，但当工作流成百上千时，就比较麻烦了。

### Anything else

_No response_

### Version

3.2.x

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by llllkid**
…h loses the flow relationship

## Purpose of the pull request

Fix #16900

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

## Pull Request Notice
[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkflowDefinitionServiceImpl.java [hunk 1]
```diff
@@ -2122,6 +2124,19 @@ protected void doBatchOperateWorkflowDefinition(User loginUser,
...
                         taskDefinitionLog.setCode(taskCode);
+                        if (TaskTypeUtils.isSwitchTask(taskDefinitionLog.getTaskType())){
+                            final String taskParams = taskDefinitionLog.getTaskParams();
+                            final SwitchParameters switchParameters = JSONUtils.parseObject(taskParams, SwitchParameters.class);
+                            if (switchParameters == null) {
+                                throw new IllegalArgumentException("Switch task params: " + taskParams + " is invalid.");
+                            }
+                            List<SwitchResultVo> dependTaskList = switchParameters.getSwitchResult().getDependTaskList();
+                            dependTaskList.forEach(switchResultVo -> {
+                                switchResultVo.setNextNode(taskCodeMap.get(switchResultVo.getNextNode()));
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkflowDefinitionServiceImpl.java:2136]: We also need to change the default branch in `SwitchResult`.

**Review by ruanwenjun — CHANGES_REQUESTED** [review]

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkflowDefinitionServiceImpl.java (hunk 1 above)
Comment (llllkid) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkflowDefinitionServiceImpl.java:2136]: > 我们还需要更改默认分支`SwitchResult`。

done

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkflowDefinitionServiceImpl.java [hunk 2]
```diff
@@ -2122,6 +2124,25 @@ protected void doBatchOperateWorkflowDefinition(User loginUser,
                         long taskCode = CodeGenerateUtils.genCode();
                         taskCodeMap.put(taskDefinitionLog.getCode(), taskCode);
                         taskDefinitionLog.setCode(taskCode);
+                        if (TaskTypeUtils.isSwitchTask(taskDefinitionLog.getTaskType())){
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkflowDefinitionServiceImpl.java]: Please add some UT.

Comment (llllkid) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkflowDefinitionServiceImpl.java]: > 请添加一些 UT。

👌，I will create the Test

**Comment by HEEKDragonOne** [comment]
没有处理掉，工作流还是不能复制逻辑节点的关联关系

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkflowDefinitionServiceImpl.java (hunk 2 above)
Comment (llllkid) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkflowDefinitionServiceImpl.java]: > Please add some UT.

done

**Comment by SbloodyS** [comment]
> 没有处理掉，工作流还是不能复制逻辑节点的关联关系

Please using english to describe.

**Comment by HEEKDragonOne** [comment]
> > 没有处理掉，工作流还是不能复制逻辑节点的关联关系
> 
> Please using english to describe.

Workflow still cannot copy the association relationship of logical nodes

**Comment by llllkid** [comment]
> > > 没有处理掉，工作流还是不能复制逻辑节点的关联关系
> > 
> > 
> > Please using english to describe.
> 
> Workflow still cannot copy the association relationship of logical nodes

Hi，what is the version of your dolphinScheduler? 
I fixed in the dev branch, I will try to upload a gif to prove that the fix was useful.
If there is nothing wrong with the gif, then you can see the results after the fix.
switch-gif-2

**Comment by HEEKDragonOne** [comment]
> > > > 没有处理掉，工作流还是不能复制逻辑节点的关联关系
> > > 
> > > 
> > > Please using english to describe.
> > 
> > 
> > Workflow still cannot copy the association relationship of logical nodes
> 
> Hi，what is the version of your dolphinScheduler? I fixed in the dev branch, I will try to upload a gif to prove that the fix was useful. If there is nothing wrong with the gif, then you can see the results after the fix. switch-gif-2 [ switch-gif-2 ](https://private-user-images.githubusercontent.com/38174237/398511762-81e6163e-5197-42f4-bf4c-5ffb547e928a.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMDIxNTgsIm5iZiI6MTczNTEwMTg1OCwicGF0aCI6Ii8zODE3NDIzNy8zOTg1MTE3NjItODFlNjE2M2UtNTE5Ny00MmY0LWJmNGMtNWZmYjU0N2U5MjhhLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI1VDA0NDQxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxZjMyZGVmMjJlOTNkOWZiM2ZlMjIyNTVkMmE1ZjE5YzFhYTM3MDBmNjc4ZDhlNWQ2MDdiMzkxYWE1ZmY4ZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.KOS5WYuKHgHIN2ZB4n0PlQbDpvftuc7EpHWkgMfdrNg) [ ](https://private-user-images.githubusercontent.com/38174237/398511762-81e6163e-5197-42f4-bf4c-5ffb547e928a.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMDIxNTgsIm5iZiI6MTczNTEwMTg1OCwicGF0aCI6Ii8zODE3NDIzNy8zOTg1MTE3NjItODFlNjE2M2UtNTE5Ny00MmY0LWJmNGMtNWZmYjU0N2U5MjhhLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI1VDA0NDQxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxZjMyZGVmMjJlOTNkOWZiM2ZlMjIyNTVkMmE1ZjE5YzFhYTM3MDBmNjc4ZDhlNWQ2MDdiMzkxYWE1ZmY4ZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.KOS5WYuKHgHIN2ZB4n0PlQbDpvftuc7EpHWkgMfdrNg)

ok

**Comment by llllkid** [comment]
@SbloodyS 
hi，this is my first time contribut。
i see the style check did’t pass, should I change the code and commit it again?

**Comment by SbloodyS** [comment]
You can run `mvn spotless:apply` to fix these violations. @JackIllkid

**Comment by llllkid** [comment]
> 您可以运行`mvn spotless:apply`来修复这些违规行为。@JackIllkid

@SbloodyS done, Can you check it again please.

**Comment by llllkid** [comment]
@SbloodyS @ruanwenjun PTAL

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkflowDefinitionServiceImpl.java [hunk 3]
```diff
@@ -2122,6 +2124,26 @@ protected void doBatchOperateWorkflowDefinition(User loginUser,
...
+                            if (switchParameters == null) {
+                                throw new IllegalArgumentException(
+                                        "Switch task params: " + taskParams + " is invalid.");
+                            }
+                            SwitchParameters.SwitchResult switchResult = switchParameters.getSwitchResult();
+                            List<SwitchResultVo> dependTaskList = switchResult.getDependTaskList();
+                            dependTaskList.forEach(switchResultVo -> {
+                                switchResultVo.setNextNode(taskCodeMap.get(switchResultVo.getNextNode()));
+                            });
+                            switchResult.setDependTaskList(dependTaskList);
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkflowDefinitionServiceImpl.java]:
```suggestion
                            switchResult.getDependTaskList().forEach(switchResultVo -> {
                                switchResultVo.setNextNode(taskCodeMap.get(switchResultVo.getNextNode()));
                            });
```

**Review by ruanwenjun — COMMENTED** [review]
LGTM, one nit.

**Comment by llllkid** [comment]
> LGTM, one nit.

good better best

**Comment by llllkid** [comment]
@caishunfeng PTAL

**Review by ruanwenjun — DISMISSED** [review]

**Review by Gallardot — APPROVED** [review]

**Review by ruanwenjun — APPROVED** [review]

**Comment by ruanwenjun** [comment]
Merged into dev, this will be released at 3.3.0, thanks for your PR.

**Comment by HEEKDragonOne** [comment]
#16934 When copying workflows containing switch nodes, if created in different orders, the associated relationships will also be lost

**Pull request merged**